lib/Bio/EnsEMBL/DataCheck/Test/DataCheck.pm

```diff
@@ -459,4 +460,72 @@ sub has_data {
   is_rows_zero($dbc, $sql, $test_name, $diag_msg);  
 }
 
+=head2 Testing one-to-many relationships  
+
+=over 4
+
+=item B<is_one_to_many>
+
+is_one_to_many($dbc, $table, $column, $test_name, $constraint);
+
+Tests that each C<$column> member is present in the table more than once.
+If all the rows have a count>1, the test will pass. C<$constraint> is an optional
+SQL where clause.
+
+=back
+
+=cut
+
+sub is_one_to_many {
+  my ($dbc, $table, $column, $test_name, $constraint) = @_;
+  
+  if (!defined $constraint) {
+    $constraint = "";
+  }
+  elsif ($constraint !~ /WHERE/) {
+    $constraint = "WHERE $constraint";
+  }
+  
+  my $sql = qq/
+    SELECT $column 
+    FROM $table
+    $constraint
+    GROUP BY $column 
+    HAVING COUNT(*) = 1
+  /;
+  
+  is_rows_zero($dbc, $sql, $test_name);
+}
+
+=head2 Testing for orphan IDs  
+
+=over 4
+
+=item B<has_no_orphans>
+
+has_no_orphans($dbc, $table1, $column1, $table2, $column2, $test_name);
+
+Tests that each C<$column1> of C<$table1> is present in C<$table2> as 
+C<$column2>. As long as C<$column1> of C<$table1> exists in C<$column2> 
+from C<$table2>, the test will pass. 
+
+=back
+
+=cut
+
+sub has_no_orphans {
+  my ($dbc, $table1, $column1, $table2, $column2, $test_name) = @_;
+
+  my $sql = qq/
+    SELECT * FROM 
+      $table1 
+    WHERE $column1 NOT IN (
+      SELECT DISTINCT $column2 
+        FROM $table2
+      )
+  /;
```

Comment:
I think it is more efficient in large data to use the NOT IN(subquery). In smaller data there is probably very little difference between the LEFT JOIN and the NOT IN (subquery). I can change it to a LEFT JOIN if you think it would be safer/preferable for other users?

Reply:
I actually think that the NOT IN version is less efficient, because MySQL has to build the list of values from the subquery and index it. With a LEFT JOIN, it can use the index that already exists.

But more importantly, now that I look back at this function, I realise it is essentially doing a foreign-key check: `species_set_tag.species_set_id` must point at a valid `species_set_header.species_set_id`. In the HCs it should have been in `compara.ForeignKeyMasterTables` instead of `compara.CheckSpeciesSetTable`. There is already a `fk` function in `Test::DataCheck` that does this.

Reply:
I was reading a benchmarking article comparing `NOT IN` with `LEFT JOIN` for several different datasets, which showed significant results for `NOT IN` being much faster.  Anyway, so remove this function altogether? And should I drop any DCs that would use this function? Such as `MemberProductionCounts` (I haven't pushed this one yet) as well as `CheckSpeciesSetTable`?

Reply:
Interesting ! Can you share the article please ?

Reply:
I would like to see how it's used in `MemberProductionCounts` first, but so far I think that `has_no_orphans` has the same role as `fk`, so one should be removed in favour of the other. The choice should be made on the basis of performance and flexibility of the method.

In terms of DC, don't drop any DC but record the usage of this function. When the foreign-keys DCs are introduced, let's check whether those function calls can move over

Reply:
[`NOT IN` is more efficient](https://planet.mysql.com/entry/?id=24888)
[`NOT IN` and `LEFT JOIN` are equally efficient](https://explainextended.com/2009/09/18/not-in-vs-not-exists-vs-left-join-is-null-mysql/)
The way MySQL optimised these queries depends on which MySQL version, but to summarise, with indexing `NOT IN` seems to be faster in general, but it is unable to handle `NULL`s, although I guess in our case, if using it for foreign keys, we don't expect any `NULL`s?
If the `fk` method is to be kept, I would like to make some other changes to it though. It is a little deceiving and not very flexible. You cannot provide a `test_name` without providing a `constraint`, although `constraint` is supposed to be optional. You currently have to provide either just a `constraint` and allow the functions default `test_name`, provide both `constraint` and `test_name` or provide neither one. It would be nice not to have to provide a `constraint` and provide just the `test_name`. I have switched the methods in this pull request to use the `fk` function instead of `has_no_orphans` and allowed the default `test_name` provided by the function for now. I have removed `has_no_orphans`, but can put it back at a later date if decided it is worth keeping additionally, or to replace the `fk` function. This way I can provide the tests with just the function to keep (`is_one_to_many`) until we know for sure. Does this sound reasonable?

Reply:
Tests on ensembl_compara_99: (query cache cleared before and between each query)
`SELECT * FROM gene_member_hom_stats WHERE gene_member_id NOT IN (SELECT DISTINCT gene_member_id FROM gene_member)
real    0m20.178s
user    0m0.054s
sys     0m0.069s

SELECT * FROM gene_member_hom_stats t1 LEFT JOIN gene_member t2 ON t1.gene_member_id = t2.gene_member_id WHERE t1.gene_member_id IS NOT NULL AND t2.gene_member_id IS NULL
real    0m27.718s
user    0m0.050s
sys     0m0.061s`

Reply:
1. Many thanks for the links, they are very interesting ! Yes since in this case NULL is not possible, LEFT JOIN is still fast enough. Can you create an "Housekeeping" Jira ticket in ENSCOMPARASW to review our usage our LEFT JOIN, with all the information you've given here ?
2. Regarding `fk`, you can in fact provide a test name but no constraint, simply by passing _undef_ to the latter. In our case, I actually think the default name is not too different from the one you wanted to give anyway.
3. About rewriting `fk`, that wouldn't work out of the box because the current $constraint parameter can target both tables (see Checks/ForeignKeys.pm line 156), which would not be possible with a NOT IN or NOT EXISTS. Since the best query depends on the presence of NULL, and LEFT JOIN works fine when NULL is not allowed, `fk` could remain as it is, but I can see the benefit of having an alternative implementation that supports NULL being a valid value of `t2.$col2`. I would choose a more explicit name than `has_no_orphans`, like `fk_null_allowed`, though. @james-monkeyshines , what's your opinion ?
4. Yes please go ahead adding a test for `is_one_to_many`

Reply:
I will create a ticket :+1: 
That is good to know about passing `undef`, I wasn't aware. I take it all back about the `fk()` function! And I agree about the wording of the `$test_name`, it is very similar to the one I would use, so it works well in this occasion.